python/ray/autoscaler/autoscaler.py

```diff
@@ -241,12 +242,14 @@ def get_resource_usage(self):
                     if frac > max_frac:
                         max_frac = frac
             nodes_used += max_frac
-
-        # This guards against the edge case where you have a small head node
-        # and large worker node, and the worker is 100% busy, but the head node
-        # is idle since it is too small to run any tasks.
-        if has_saturated_node and nodes_used < num_nodes:
-            nodes_used = min(nodes_used + 1.0, num_nodes)
+            if max_frac > 0:
+                num_nonidle += 1
+
+        # If any nodes have a queue buildup, assume all non-idle nodes are 100%
```

Comment:
This seems very aggressive, for example in a case where we somehow have nearly the whole cluster idle but a single node with a queue (could happen if there were many short-lived tasks scheduled on other nodes?). Maybe we should make it a bit more of a heuristic instead of this hard rule. Seems like a fine fix for now, though.

Reply:
This won't trigger an up-scaling event unless there is at least "some" load on almost every node. So the case you mentioned is probably OK.

But yeah agreed we could be smarter, perhaps by looking at the queue size and extrapolating the load.